File: src/calculator/expo/cards/effects/hopeful-refugees.ts
import type {
  DefaultCalculatorReturnType,
  GenericCalculatorParams,
} from '@nanosh/types/generic'
import GetRandomNumber from '@nanosh/utils/getRandomNumber'
import seedrandom from 'seedrandom'

interface HopefulRefugeesParams
  extends Pick<GenericCalculatorParams, 'state' | 'gameID' | 'invokeTime'> {}

const HOPEFUL_REFUGEES_MIN_CIVITATES = 3
const HOPEFUL_REFUGEES_MAX_CIVITATES = 6

export default function HopefulRefugees({
  state,
  gameID,
  invokeTime,
}: HopefulRefugeesParams): DefaultCalculatorReturnType {
  const stateCopy = structuredClone(state)

  const prng = seedrandom(`${gameID}-${invokeTime}`)
  const civitatesAdded = GetRandomNumber(
    HOPEFUL_REFUGEES_MIN_CIVITATES,
    HOPEFUL_REFUGEES_MAX_CIVITATES,
    prng,
  )

  stateCopy.ship.civitates += civitatesAdded

  return [stateCopy, null]
}

File: src/nakama/gamelogic/calculator/expedition/init/ally/hopeful_refugees.go
package ally

import (
	expeditionErrors "nanosh/nakama-modules/gamelogic/calculator/expedition"
	"nanosh/nakama-modules/gamelogic/types"
	"nanosh/nakama-modules/gamelogic/utils/dice"
	"nanosh/nakama-modules/gamelogic/utils/random"
)

// HopefulRefugees - Gain 3-6 civitates
func HopefulRefugees(game *types.Game, gameID string, invokeTime int64, username string) (*types.Game, error) {
	// Check expedition exists
	if game.Expedition == nil {
		return nil, expeditionErrors.ErrNoActiveExpedition
	}

	// Create deterministic PRNG using standard pattern
	prng := random.CreateSeededPRNG(gameID, invokeTime, username)

	// Roll for civitates (3-6) using dice utils
	civitatesGain := dice.RollRange(prng, 3, 6)

	// Add civitates to ship
	game.Ship.Civitates += civitatesGain

	return game, nil
}
